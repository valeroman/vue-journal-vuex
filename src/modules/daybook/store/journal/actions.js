// Pueden ser acciones asincronas que pueden llamar a una mutacion
// export const myAction = async({ commit }) => {

import journalApi from "../../../../api/journalApi"

// }

export const loadEntries = async({ commit }) => {

    const { data } = await journalApi.get('/entries.json')

    if (!data) {
        commit('setEntries', [])
        return
    }

    // console.log({ data })

    const entries = []
    for (let id of Object.keys(data)) {
        entries.push({
            id,
            ...data[id] // el valor al que apunta el id
        })
    }
    // console.log(entries)
    commit('setEntries', entries)

}

export const updateEntry = async({ commit }, entry) => {

    // Extraer solo lo que necesita
    const { date, text, picture } = entry
    const dataToSave = { date, text, picture }

    // await journalApi.put(PATH .json, dataToSave)
    await journalApi.put(`/entries/${ entry.id }.json`, dataToSave)

    dataToSave.id = entry.id

    // commit de una mutacion => updateEntry
    commit('updateEntry', {...dataToSave })
}

export const createEntry = async({ commit }, entry) => {

    const { date, text, picture } = entry
    const dataToSave = { date, text, picture }

    const { data } = await journalApi.post('/entries.json', dataToSave)

    dataToSave.id = data.name

    commit('addEntry', dataToSave)

    return data.name
}

export const deleteEntry = async({ commit }, id) => {

    await journalApi.delete(`/entries/${ id }.json`)

    commit('deleteEntry', id)

    return id
}