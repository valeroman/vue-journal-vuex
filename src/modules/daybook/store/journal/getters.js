// Se puede traer informacion del state
// export const myGetter = (state) => {
//  return state;
// }

export const getEntriesByTerm = ( /* state */ ) => {

}

export const getEntryById = ( /* state */ ) => {

}