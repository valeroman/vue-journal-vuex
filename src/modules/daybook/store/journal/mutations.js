// Son sincronas y hacen la modificacion del state
// export const myMutation = (state) => {

// }

export const setEntries = ( /* state */ ) => {

}

export const updateEntries = ( /* state */ ) => {

}

export const addEntries = ( /* state */ ) => {

}