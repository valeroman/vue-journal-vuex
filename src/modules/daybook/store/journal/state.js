// El state es reactivo y cuando se cambia, se le notifica a todos los componentes que esten escuchando
export default () => ({
    isLoading: true,
    entries: []
})