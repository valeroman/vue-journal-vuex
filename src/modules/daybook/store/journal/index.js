import * as actions from './actions'
import * as getters from './getters'
import * as mutations from './mutations'
import state from './state'

const journalModule = {
    namespaced: true,
    actions,
    mutations,
    getters,
    state
}

export default journalModule