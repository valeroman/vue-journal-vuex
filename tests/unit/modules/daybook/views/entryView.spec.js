import journal from '@/modules/daybook/store/journal'
import EntryView from '@/modules/daybook/views/EntryView'
import { shallowMount } from '@vue/test-utils'
import Swal from 'sweetalert2'
import { createStore } from 'vuex'
import { journalState } from '../../../mock-data/test-journal-state'


const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: {...initialState }
            }
        }
    })

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))


describe('Pruebas en el EntryView', () => {

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-MqVkQ82xiM6NdioDkv9-'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('debe de sacar al usuario por que el id no existe', () => {

        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'Este ID no existe en el STORE'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]

            }
        })

        expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' })
    })

    test('debe de mkostrar la entrada correctamente', () => {

        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()
    })

    test('debe de borrar la entrada y salir', (done) => {

        Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }))

        wrapper.find('.btn-danger').trigger('click')

        expect(Swal.fire).toHaveBeenCalledWith({
            title: 'Estas seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si estoy seguro'
        })

        setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-MqVkQ82xiM6NdioDkv9-')
            expect(mockRouter.push).toHaveBeenCalled()
            done()
        }, 1)
    })

})