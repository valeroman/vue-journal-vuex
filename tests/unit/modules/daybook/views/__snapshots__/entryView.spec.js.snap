// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mkostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">9</span><span class="mx-1 fs-3">Diciembre</span><span class="mx-2 fs-4 fw-light">2021, Jueves</span></div>
  <div><input type="file" accept="image/png, image/jpg, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="QuÃ© paso hoy?">
            </textarea></div>
<img src="https://res.cloudinary.com/dddudqtmm/image/upload/v1639107994/glowl4bb8nwhmxpqsvp0.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
